### PR TITLE
Check empty features

### PR DESCRIPTION
**Pull request by saracastel**
# Description
Checks if there are any features with empty gene symbol both in the features file and in the count matrix, and remove them if < 0.1% of the total number of features

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1907

#### Link to staging deployment URL (or set N/A)
https://ui-sara-pipeline233.scp-staging.biomage.net/

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [x] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 1]
```diff
@@ -80,6 +80,11 @@ read_10x_files <- function(config, input_dir) {
       counts <- counts[[slot]]
     }
 
+    # Check if there are any rows with empty gene symbol in the count matrix and remove them if < 0.1% of the total number of features
+    if (length(which(rownames(counts) == "")) != 0 & length(which(rownames(counts) == ""))/nrow(counts) < 0.001) {
+      counts <- counts[!(rownames(counts) == ""),]
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:85]: given that you use this three times, I would extract the `which(rownames(counts) == "")` as a variable. this way, you perform the string comparation only once. 

Also, is the first condition really necessary? if the second one is true, the first one cannot be false.

Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:85]:
```suggestion
  unnamed_genes <- which(rownames(counts) == "")
  
  if (length(unnamed_genes) / nrow(counts) < 0.001) {
    counts <- counts[!unnamed_genes,]
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 2]
```diff
@@ -92,6 +97,11 @@ read_10x_files <- function(config, input_dir) {
       )
     )
 
+    # Check if there are any rows with empty gene symbol in the feature file and remove them if < 0.1% of the total number of features
+    if (length(which(annot[,1] == "")) != 0 & length(which(annot[,1] == ""))/nrow(annot) < 0.001) {
+      annot <- annot[!(annot[,1] == ""),]
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:102]: same idea here

**Comment by gerbeldo** [comment]
also, you should write some tests :)

File: pipeline-runner/R/gem2s-2-load_user_files.R (hunk 1 above)
Comment (gerbeldo) [reply pipeline-runner/R/gem2s-2-load_user_files.R:85]:
```suggestion
  unnamed_genes <- which(rownames(counts) == "")
	  
	  if (length(unnamed_genes) != 0 & length(unnamed_genes) / nrow(counts) < 0.001) {
	    counts <- counts[!unnamed_genes, ]
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 3]
```diff
@@ -92,6 +99,13 @@ read_10x_files <- function(config, input_dir) {
       )
     )
 
+    # Check if there are any rows with empty gene symbol in the feature file and remove them if < 0.1% of the total number of features
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:102]: remove the comment, since the hyperparam no longer applies

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 4]
```diff
@@ -80,6 +80,13 @@ read_10x_files <- function(config, input_dir) {
       counts <- counts[[slot]]
     }
 
+    # Check if there are any rows with empty gene symbol in the count matrix and remove them if < 0.1% of the total number of features
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:83]: fix comment (0.1% not valid anymore)

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R [hunk 5]
```diff
@@ -443,3 +443,65 @@ test_that("parse_rhapsody_matrix uses RSEC if DBEC corrected counts are missing"
...
+  outdir <- tempdir()
+  sample <- "sample_a"
+  sample_dir <- file.path(outdir, sample)
+  dir.create(sample_dir)
+
+  mock_cellranger_files(counts, features, sample_dir)
+
+  prev_out <- list(config = list(samples = sample, input = list(type = "10x")))
+
+  out <- pipeline::load_user_files(NULL, NULL, prev_out, outdir)$output
```
Comment (gerbeldo) [inline pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:467]: This is not necessary, it's an exported function in the pipeline

```suggestion
  out <- load_user_files(NULL, NULL, prev_out, outdir)$output
```

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R [hunk 6]
```diff
@@ -443,3 +443,65 @@ test_that("parse_rhapsody_matrix uses RSEC if DBEC corrected counts are missing"
...
+
+  counts_list <- out$counts_list
+  annot <- out$annot
+
+  expect_true(length(which(rownames(counts) == "")) > 0 & length(which(rownames(counts_list[[1]]) == "")) == 0)
```
Comment (gerbeldo) [inline pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:472]: I would remove that condition, given that it's fixed by us. Unit tests are about the functionality of the function being tested.

```suggestion
  expect_true(length(which(rownames(counts_list[[1]]) == "")) == 0)
```

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R [hunk 7]
```diff
@@ -443,3 +443,65 @@ test_that("parse_rhapsody_matrix uses RSEC if DBEC corrected counts are missing"
...
+  mock_cellranger_files(counts, features, sample_dir)
+
+  prev_out <- list(config = list(samples = sample, input = list(type = "10x")))
+
+  out <- load_user_files(NULL, NULL, prev_out, outdir)$output
+
+  counts_list <- out$counts_list
+  annot <- out$annot
+
+  expect_true(length(which(rownames(counts) == "")) == 0 & length(which(rownames(counts_list[[1]]) == "")) == 0)
```
Comment (gerbeldo) [inline pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:500]:
```suggestion
  expect_true(length(which(rownames(counts_list[[1]]) == "")) == 0)
```

**Review by gerbeldo — APPROVED** [review]
some small comments but looks good!

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R [hunk 8]
```diff
@@ -443,3 +443,129 @@ test_that("parse_rhapsody_matrix uses RSEC if DBEC corrected counts are missing"
...
+test_that("read_10x_files removes rows with empty feature names both in count matrix and annotation if present and < 0.1%", {
+  counts <- mock_counts()[rep(seq_len(nrow(mock_counts())), each = 10), ]
+  rownames(counts)[2]=""
+  rownames(counts)[3]=".1"
+
+  features <- data.frame(
+    ensid = paste0("ENSFAKE", seq_len(nrow(counts))),
+    symbol = row.names(counts)
+  )
+  features[2:3,1:2] = ""
```
Comment (gerbeldo) [inline pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:457]: we use arrows for assignment!
```suggestion
  features[2:3,1:2] <- ""
```

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R [hunk 9]
```diff
@@ -443,3 +443,129 @@ test_that("parse_rhapsody_matrix uses RSEC if DBEC corrected counts are missing"
...
+
+test_that("read_10x_files removes single row with empty feature names both in count matrix and annotation if present and < 0.1%", {
+  counts <- mock_counts()[rep(seq_len(nrow(mock_counts())), each = 10), ]
+  rownames(counts)[2]=""
+
+  features <- data.frame(
+    ensid = paste0("ENSFAKE", seq_len(nrow(counts))),
+    symbol = row.names(counts)
+  )
+  features[2,1:2] = ""
```
Comment (gerbeldo) [inline pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:488]:
```suggestion
  features[2,1:2] <- ""
```

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R [hunk 10]
```diff
@@ -443,3 +443,129 @@ test_that("parse_rhapsody_matrix uses RSEC if DBEC corrected counts are missing"
...
+  expect_true(length(which(annot[,1] == "")) == 0)
+
+  unlink(sample_dir, recursive = TRUE)
+})
+
+
+test_that("read_10x_files doesn't remove any rows with empty feature names both in count matrix and annotation if present and >= 0.1%", {
+  counts <- mock_counts()
+  rownames(counts)[2]=""
+  rownames(counts)[3]=".1"
```
Comment (gerbeldo) [inline pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:514]:
```suggestion
  rownames(counts)[2] <- ""
  rownames(counts)[3] <- ".1"
```

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R [hunk 11]
```diff
@@ -443,3 +443,129 @@ test_that("parse_rhapsody_matrix uses RSEC if DBEC corrected counts are missing"
...
+
+  features <- data.frame(
+    ensid = paste0("ENSFAKE", seq_len(nrow(counts))),
+    symbol = row.names(counts)
+  )
+  features[2:3,1:2] = ""
```
Comment (gerbeldo) [inline pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:520]:
```suggestion
  features[2:3,1:2] <- ""
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 12]
```diff
@@ -92,6 +100,13 @@ read_10x_files <- function(config, input_dir) {
       )
     )
 
+    # Check if there are any rows with empty gene symbol in the feature file and remove them if < 0.1% of the total number of features
+    unnamed_ids <- which(annot[,1] == "")
+
+    if (length(unnamed_ids) != 0 & length(unnamed_ids)/nrow(annot) < 0.001) {
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:106]:
```suggestion
    if (length(unnamed_ids) != 0 & length(unnamed_ids) / nrow(annot) < 0.001) {
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 13]
```diff
@@ -80,6 +80,14 @@ read_10x_files <- function(config, input_dir) {
       counts <- counts[[slot]]
     }
 
+    # Check if there are any rows with empty gene symbol in the count matrix and remove them if < 0.1% of the total number of features
+    # if more than 1 empty gene is present in the feature file, the count matrix will have the first rowname empty, while the following will be ".1", ".2". etc...
+    unnamed_genes <- c(which(rownames(counts) == ""),grep("^\\.[0-9]+", rownames(counts)))
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:85]:
```suggestion
    unnamed_genes <- c(which(rownames(counts) == ""), grep("^\\.[0-9]+", rownames(counts)))
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 14]
```diff
@@ -80,6 +80,14 @@ read_10x_files <- function(config, input_dir) {
       counts <- counts[[slot]]
     }
 
+    # Check if there are any rows with empty gene symbol in the count matrix and remove them if < 0.1% of the total number of features
+    # if more than 1 empty gene is present in the feature file, the count matrix will have the first rowname empty, while the following will be ".1", ".2". etc...
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:84]:
```suggestion
    # if more than 1 empty gene is present in the feature file,
    # the count matrix will have the first rowname empty, while the 
    # following will be ".1", ".2". etc...
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 15]
```diff
@@ -80,6 +80,14 @@ read_10x_files <- function(config, input_dir) {
       counts <- counts[[slot]]
     }
 
+    # Check if there are any rows with empty gene symbol in the count matrix and remove them if < 0.1% of the total number of features
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:83]:
```suggestion
    # Check if there are any rows with empty gene symbol in the count
    # matrix and remove them if < 0.1% of the total number of features
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 16]
```diff
@@ -80,6 +80,14 @@ read_10x_files <- function(config, input_dir) {
...
+    # if more than 1 empty gene is present in the feature file, the count matrix will have the first rowname empty, while the following will be ".1", ".2". etc...
+    unnamed_genes <- c(which(rownames(counts) == ""),grep("^\\.[0-9]+", rownames(counts)))
+
+    if (length(unnamed_genes) != 0 & length(unnamed_genes)/nrow(counts) < 0.001) {
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:87]:
```suggestion
    if (length(unnamed_genes) != 0 & length(unnamed_genes) / nrow(counts) < 0.001) {
```

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R [hunk 17]
```diff
@@ -443,3 +443,129 @@ test_that("parse_rhapsody_matrix uses RSEC if DBEC corrected counts are missing"
 
   expect_equal(values, expected_values)
 })
+
+
+test_that("read_10x_files removes rows with empty feature names both in count matrix and annotation if present and < 0.1%", {
+  counts <- mock_counts()[rep(seq_len(nrow(mock_counts())), each = 10), ]
```
Comment (gerbeldo) [inline pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:449]: why is this necessary? maybe add a comment

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R [hunk 18]
```diff
@@ -443,3 +443,129 @@ test_that("parse_rhapsody_matrix uses RSEC if DBEC corrected counts are missing"
...
+  rownames(counts)[2]=""
+  rownames(counts)[3]=".1"
```
Comment (gerbeldo) [inline pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:451]:
```suggestion
  rownames(counts)[2] <- ""
  rownames(counts)[3] <- ".1"
```

**Review by gerbeldo — APPROVED** [review]
some styling issues, so you should run styler on these files. also, careful on using "=" for assignments!

File: pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R (hunk 17 above)
Comment (saracastel) [reply pipeline-runner/tests/testthat/test-gem2s-2-load_user_files.R:457]: this is to increase the size of counts matrix, to be able to mock a matrix with < 0.1% of empty features. I'll add a comment

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 19]
```diff
@@ -80,6 +80,17 @@ read_10x_files <- function(config, input_dir) {
       counts <- counts[[slot]]
     }
 
+    # Check if there are any rows with empty gene symbol in the count
+    # matrix and remove them if < 0.1% of the total number of features
+    # if more than 1 empty gene is present in the feature file,
+    # the count matrix will have the first rowname empty, while the 
+    # following will be ".1", ".2". etc...
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:87]:
```suggestion
    # Check if there are any rows with empty gene symbols in the count
    # matrix and remove them if < 0.1% of the total number of features.
    # If more than 1 empty gene is present in the features file,
    # the count matrix will have the first rowname empty, while the 
    # following will be ".1", ".2"... etc. (because make.unique)
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 20]
```diff
@@ -92,6 +103,13 @@ read_10x_files <- function(config, input_dir) {
       )
     )
 
+    # Check if there are any rows with empty gene symbol in the feature file and remove them if < 0.1% of the total number of features
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:106]:
```suggestion
    # Check if there are any rows with empty gene symbol in the features file
    # and remove them if < 0.1% of the total number of features
```

File: pipeline-runner/R/gem2s-2-load_user_files.R (hunk 19 above)
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:87]:
```suggestion
    # Check existence of empty gene symbols in count matrix' rownames.
    # If there are more than one the first will be empty, while the 
    # following will be ".1", ".2"... because Seurat runs make.unique
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 21]
```diff
@@ -80,6 +80,17 @@ read_10x_files <- function(config, input_dir) {
       counts <- counts[[slot]]
     }
 
+    # Check if there are any rows with empty gene symbol in the count
+    # matrix and remove them if < 0.1% of the total number of features
+    # if more than 1 empty gene is present in the feature file,
+    # the count matrix will have the first rowname empty, while the 
+    # following will be ".1", ".2". etc...
+    unnamed_genes <- c(which(rownames(counts) == ""), grep("^\\.[0-9]+", rownames(counts)))
+
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:89]:
```suggestion
    # remove rows with empty names if < 0.1% of the total features.
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 22]
```diff
@@ -80,6 +80,21 @@ read_10x_files <- function(config, input_dir) {
...
+    unnamed_genes <- c(which(rownames(counts) == ""), grep("^\\.[0-9]+", rownames(counts)))
+    # remove rows with empty names if < 0.1% of the total features.
+    if (length(unnamed_genes) != 0 & length(unnamed_genes) / nrow(counts) < 0.001) {
+      counts <- counts[-unnamed_genes,]
+       message(
+        sprintf(
+          "*** Removed %s rows with empty gene symbol from count matrix of sample %s",
+          length(unnamed_genes), sample
+        )
+      )
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:95]:
```suggestion
       message(
        sprintf(
          "Removed %s rows with empty gene symbol from count matrix of sample %s",
          length(unnamed_genes), sample
        )
      )
```

File: pipeline-runner/R/gem2s-2-load_user_files.R [hunk 23]
```diff
@@ -92,6 +107,20 @@ read_10x_files <- function(config, input_dir) {
...
 
+    # Check if there are any rows with empty gene symbol in the features file
+    # and remove them if < 0.1% of the total number of features
+    unnamed_ids <- which(annot[,1] == "")
+
+    if (length(unnamed_ids) != 0 & length(unnamed_ids) / nrow(annot) < 0.001) {
+      annot <- annot[-unnamed_ids,]
+      message(
+        sprintf(
+          "*** Removed %s rows with empty gene symbol from annot of sample %s",
```
Comment (gerbeldo) [inline pipeline-runner/R/gem2s-2-load_user_files.R:118]:
```suggestion
          "Removed %s rows with empty gene symbol from annot of sample %s",
```

**Comment by alexvpickering** [comment]
@saracastel @gerbeldo some questions about this one:

- Why is the solution to remove these genes and why does it matter how many there are (threshold)? 
- Why not check to see if the missing ENSEMBL IDs have SYMBOLS and use those instead?
- Shouldn't we either delete any genes that have no annotation or throw a helpful error (why would we keep genes that can not be identified)? 

Thanks!

**Comment by saracastel** [comment]
These are very good points, thanks @alexvpickering! We thought that the best way to proceed is to address them inside this [other PR](https://github.com/hms-dbmi-cellenics/pipeline/pull/241), given that there are some conflicts occurring between these two branches.
